File: Question 0046 - Permutations/src/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private void solve(int[] nums,List<Integer> permutations, List<List<Integer>> ans)
    {
        if(permutations.size() == nums.length)
        {
            ans.add(new ArrayList<>(permutations));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if(permutations.contains(nums[i])) continue;

            permutations.add(nums[i]);
            solve(nums,permutations, ans);
            permutations.remove(permutations.size()-1);
        }
    }
    public List<List<Integer>> permute(int[] nums)
    {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> permutations = new ArrayList<>();
        solve(nums,permutations,  ans);
        return ans;
    }
}

File: Question 0046 - Permutations/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

void solve(vector<int>& nums, vector<int>& curr, vector<vector<int>>& ans)
{
    if(curr.size() == nums.size())
    {
        ans.push_back(curr);
        return;
    }

    for(int i = 0; i < nums.size(); i++)
    {
        if(find(curr.begin(), curr.end(), nums[i]) != curr.end()) continue;

        curr.push_back(nums[i]);
        solve(nums, curr, ans);
        curr.pop_back();
    }
}

class Solution 
{
public:
    vector<vector<int>> permute(vector<int>& nums) 
    {
        vector<int> curr;
        vector<vector<int>> ans;
        solve(nums, curr, ans);
        return ans;
    }
};